widgets/amdgpu_linux.lua

```diff
@@ -0,0 +1,62 @@
+-- contrib/amdgpu_linux.lua
+-- Copyright (C) 2022  Rémy Clouard <shikamaru shikamaru fr>
+--
+-- This file is part of Vicious.
+--
+-- Vicious is free software: you can redistribute it and/or modify
+-- it under the terms of the GNU General Public License as
+-- published by the Free Software Foundation, either version 2 of the
+-- License, or (at your option) any later version.
+--
+-- Vicious is distributed in the hope that it will be useful,
+-- but WITHOUT ANY WARRANTY; without even the implied warranty of
+-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+-- GNU General Public License for more details.
+--
+-- You should have received a copy of the GNU General Public License
+-- along with Vicious.  If not, see <https://www.gnu.org/licenses/>.
+local io = { open = io.open }
+local helpers = require("vicious.helpers")
+
+-- {{{ AMDGPU widget type
+return helpers.setcall(function (format, warg)
+    if not warg then return end
+
+    local amdgpu = "/sys/class/drm/"..warg.."/device"
+    local _data = {}
+    local _mem = nil
+
+    local f = io.open(amdgpu .. "/gpu_busy_percent", "r")
+    if f then
+        for line in f:lines() do
+            _data["{gpu_usage}"] = line
+        end
+        f:close()
+    else
+        _data["{gpu_usage}"] = "N/A"
+    end
+
+    if _mem == nil then
```

Comment:
My Lua is rusty now but I suppose `_mem == nil` should always be true here.

Reply:
yeah, I wanted to put that variable in a closure based on the assumption that the max vram wouldn’t change on subsequent calls, but atm it’s a local variable which means it’s called every time, so it could be optimized there, I admit I’m not a lua expert either, how would you do that, define that variable outside of the function ? In practice I don’t notice a negative performance effect but I don’t update the associated widget as often as the performance graph.

Reply:
> I wanted to put that variable in a closure based on the assumption that the max vram wouldn’t change on subsequent calls, [...], how would you do that, define that variable outside of the function ?

Yes, some other widgets also do it.